Add s2i support

kosciCZ opened the pull request:
This PR would add s2i extensibility to the image. The image would remain usable both in the usual way and as a s2i builder image.

This would fix #12

TomasTomecek commented on s2i/bin/run:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+source /files/start.sh
```
so no way to replace `start.sh`?

kosciCZ replied:
I guess we could move contents of /files/start.sh to this file. Would that be more suitable?

TomasTomecek replied:
I'll be honest: I don't know. I would probably leave it as it is and wait for some feedback whether it's useful like this or someone wants to do some absolutely crazy stuff.

kosciCZ replied:
Now that you've pointed it out, I think we can safely move & delete this file. If someone would like to make changes to the startup routine, they can use pre-init script to edit/replace /usr/local/s2i/run. 

AFAIK the reason this file is separated is that it is a leftover from when this had bunch of startup scripts, that were specific to postfix variant.

TomasTomecek commented on files/start.sh:
```diff
@@ -24,6 +24,13 @@ if [[ ! -z "${DEBUG_MODE}" ]]; then
+#execute pre-init scripts if this is an s2i executed 
+if [[ -d "${APP_DATA}/src/postfix-pre-init" ]]; then
+    echo "S2I is enabled, sourcing pre-init scripts..."
+    process_extending_files ${APP_DATA}/src/postfix-pre-init
```
I am thinking why `src/postfix-pre-init`, why not just `src/`. Is the reason is it can be extended also to `post-init` in future?

kosciCZ replied:
I was trying to follow sort of an standard set by SCL so it would be less confusing for people. See, in SCL images, they have a directory structure like this:

| Folder name       | Description                |
|-------------------|----------------------------|
| `./${IMAGE_NAME}-cfg`     | Can contain additional configuration files |
| `./${IMAGE_NAME}-pre-init`| Can contain shell scripts (`*.sh`) that are sourced before `${IMAGE_NAME}` is started|
| `./`              | Application source code |

For reference see https://github.com/sclorg/httpd-container/tree/master/2.4#usage
or https://github.com/sclorg/mongodb-container/tree/master/3.2#extending-image

While I agree there is probably no app source code, I'd still like to do it this way, for easier use.

TomasTomecek replied:
Okay, sounds good.

TomasTomecek commented:
General comment: the PR looks good, I would like to see two things:

1. A real example of the s2i extending process so people can just copy/paste it and try it out.
2. A test case for the s2i extending, probably not blocking this PR: we can create issue and work on it later.

kosciCZ commented:
Thanks for your comments! Can maybe @phracek think of suitable test case?

TomasTomecek commented:
@kosciCZ I would do something very simple, just change one parameter using `postconf`, e.g. log verbosity

kosciCZ commented:
@TomasTomecek please check the changes I made reflecting your comments, when you have a moment. I have added an example utilizing s2i extensibility and moved start.sh to s2i run script.

TomasTomecek commented on README.md:
```diff
@@ -62,6 +62,32 @@ For more information about Postfix TLS support see `http://www.postfix.org/TLS_R
+docker run -it \
+    -p 25:10025\
+    -v /var/spool/postfix:/var/spool/postfix \
+    -v /var/spool/mail:/var/spool/mail postfix
```
I think this should be
```
-v /var/spool/mail:/var/spool/mail \
```

TomasTomecek commented on README.md.template:
```diff
@@ -62,6 +62,32 @@ For more information about Postfix TLS support see `http://www.postfix.org/TLS_R
+docker run -it \
+    -p 25:10025\
+    -v /var/spool/postfix:/var/spool/postfix \
+    -v /var/spool/mail:/var/spool/mail {{ spec.envvars.name }}
```
oh right, here's the issue, that `{{ spec.envvars.name }}` should be `\`.
